File: kotlin/app/src/main/kotlin/org/example/CreateFormFailure.kt
package org.example

sealed class CreateFormFailure {
    data object InvalidUrl : CreateFormFailure()

    data object LocationHeaderNotPresent : CreateFormFailure()
    data class RequestNotExecuted(val exception: Throwable): CreateFormFailure()
    data class ErrorResponse(val code: Int, val body: String?) : CreateFormFailure()
}

File: scala/src/main/scala/com/example/CreateFormFailure.scala
package com.example

sealed trait CreateFormFailure
object CreateFormFailure {
  case object InvalidUrl extends CreateFormFailure
  case object LocationHeaderNotPresent extends CreateFormFailure
  final case class RequestNotExecuted(exception: Throwable) extends CreateFormFailure
  final case class ErrorResponse(code: Int, body: Option[String] = None) extends CreateFormFailure
}
